### PR TITLE
Fix DynRankView in ScratchSpace

### PR DESCRIPTION
**Pull request by crtrott**
This fixes using DynRankView in scratch space. Two issues are fixed:

* `shmem_size` didn't report correct size.
* ctor taking scratch space called host only function.

### CODE REVIEW

File: containers/unit_tests/TestDynRankView_TeamScratch.hpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include <Kokkos_DynRankView.hpp>
```
Comment (dalg24) [inline containers/unit_tests/TestDynRankView_TeamScratch.hpp:17]: Include what you need, `<gtest/gtest.h>`

File: containers/unit_tests/TestDynRankView_TeamScratch.hpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+namespace Test {
```
Comment (dalg24) [inline containers/unit_tests/TestDynRankView_TeamScratch.hpp:19]: Prefer anonymous namespace
```suggestion
namespace {
```

File: containers/unit_tests/CMakeLists.txt [hunk 3]
```diff
@@ -19,6 +19,7 @@ foreach(Tag Threads;Serial;OpenMP;HPX;Cuda;HIP;SYCL)
       DynViewAPI_generic
       DynViewAPI_rank12345
       DynViewAPI_rank67
+      DynRankView_TeamScratch
```
Comment (dalg24) [inline containers/unit_tests/CMakeLists.txt:22]: This naming scheme somewhat clashes with all the other existing tests in this list...

File: containers/unit_tests/TestDynRankView_TeamScratch.hpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  Kokkos::View<unsigned, execution_space, Kokkos::MemoryTraits<Kokkos::Atomic>>
+      errors("errors");
+  auto policy = policy_type(1, Kokkos::AUTO)
+                    .set_scratch_size(0, Kokkos::PerTeam(shmem_size));
+  Kokkos::parallel_for(
+      policy, KOKKOS_LAMBDA(const team_type& team) {
+        drv_type scr(team.team_scratch(0), N0, N1, N2);
+        // Control that the code ran at all
+        if (scr.rank() != 3) errors() |= 1u;
+        if (scr.extent(0) != N0) errors() |= 2u;
```
Comment (dalg24) [inline containers/unit_tests/TestDynRankView_TeamScratch.hpp:41]: Fine by me but I expected the compiler would complain about comparing signed and unsigned integers here

File: containers/unit_tests/TestDynRankView_TeamScratch.hpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        Kokkos::parallel_for(
+            Kokkos::TeamThreadMDRange(team, N0, N1, N2),
+            [=](int i, int j, int k) { scr(i, j, k) = i * 100 + j * 10 + k; });
+        team.team_barrier();
+        Kokkos::parallel_for(Kokkos::TeamThreadMDRange(team, N0, N1, N2),
+                             [=](int i, int j, int k) {
+                               if (scr(i, j, k) != i * 100 + j * 10 + k)
+                                 errors() &= 16u;
+                             });
+        errors() |= 256u;
```
Comment (dalg24) [inline containers/unit_tests/TestDynRankView_TeamScratch.hpp:53]: Checking that something did indeed execute, I see

File: containers/unit_tests/TestDynRankView_TeamScratch.hpp [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+//
+//@HEADER
+
+#include <Kokkos_DynRankView.hpp>
+
+namespace Test {
+
+void test_dyn_rank_view_team_scratch() {
+  using execution_space = TEST_EXECSPACE;
+  using memspace_type   = execution_space::scratch_memory_space;
```
Comment (dalg24) [inline containers/unit_tests/TestDynRankView_TeamScratch.hpp:23]: Silly name

File: containers/unit_tests/TestDynRankView_TeamScratch.hpp [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  using drv_type        = Kokkos::DynRankView<int, memspace_type>;
+  using policy_type     = Kokkos::TeamPolicy<execution_space>;
+  using team_type       = policy_type::member_type;
```
Comment (dalg24) [inline containers/unit_tests/TestDynRankView_TeamScratch.hpp:26]: Nitpick: Defining a whole lot of aliases that are only used once when really only the execution space is repeated.
I would drop them and template the whole thing on ExecutionSpace.
Not blocking obviously.

File: containers/unit_tests/TestDynRankView_TeamScratch.hpp [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+  int N0 = 10, N1 = 4, N2 = 3;
+  size_t shmem_size = drv_type::shmem_size(N0, N1, N2);
+  ASSERT_GE(shmem_size, N0 * N1 * N2 * sizeof(int));
```
Comment (dalg24) [inline containers/unit_tests/TestDynRankView_TeamScratch.hpp:30]: Why greater equal?  Might be worth a comment.

**Review by dalg24 — COMMENTED** [review]
Mostly minor comments.
The only one I really want to address is the why are we not checking for equality

File: containers/src/Kokkos_DynRankView.hpp [hunk 9]
```diff
@@ -846,34 +846,19 @@ class DynRankView : private View<DataType*******, Properties...> {
...
-           (arg_N2 != KOKKOS_INVALID_INDEX ? arg_N2 : 1) *
-           (arg_N3 != KOKKOS_INVALID_INDEX ? arg_N3 : 1) *
-           (arg_N4 != KOKKOS_INVALID_INDEX ? arg_N4 : 1) *
-           (arg_N5 != KOKKOS_INVALID_INDEX ? arg_N5 : 1) *
-           (arg_N6 != KOKKOS_INVALID_INDEX ? arg_N6 : 1) *
-           (arg_N7 != KOKKOS_INVALID_INDEX ? arg_N7 : 1);
+  static inline size_t shmem_size(
+      const size_t arg_N0 = 1, const size_t arg_N1 = 1, const size_t arg_N2 = 1,
+      const size_t arg_N3 = 1, const size_t arg_N4 = 1, const size_t arg_N5 = 1,
+      const size_t arg_N6 = 1, const size_t arg_N7 = KOKKOS_INVALID_INDEX) {
```
Comment (masterleinad) [inline containers/src/Kokkos_DynRankView.hpp:852]: Would you mind adding a comment saying that view_type has rank 7 so we always have to pass 7 valid arguments?

Why did you need to change this function anyway?

File: containers/unit_tests/CMakeLists.txt (hunk 3 above)
Comment (crtrott) [reply containers/unit_tests/CMakeLists.txt:22]: I would want to rename the other ones, since its confusing whether DynView means DynamicView or DynRankView

File: containers/src/Kokkos_DynRankView.hpp (hunk 9 above)
Comment (crtrott) [reply containers/src/Kokkos_DynRankView.hpp:852]: Regarding the deleted question: it didn't return bytes.

File: containers/unit_tests/TestDynRankView_TeamScratch.hpp (hunk 5 above)
Comment (tcclevenger) [reply containers/unit_tests/TestDynRankView_TeamScratch.hpp:53]: Why not `32u`?

File: containers/unit_tests/TestDynRankView_TeamScratch.hpp [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        if (scr.extent(1) != N1) errors() |= 4u;
+        if (scr.extent(2) != N2) errors() |= 8u;
+        Kokkos::parallel_for(
+            Kokkos::TeamThreadMDRange(team, N0, N1, N2),
+            [=](int i, int j, int k) { scr(i, j, k) = i * 100 + j * 10 + k; });
+        team.team_barrier();
+        Kokkos::parallel_for(Kokkos::TeamThreadMDRange(team, N0, N1, N2),
+                             [=](int i, int j, int k) {
+                               if (scr(i, j, k) != i * 100 + j * 10 + k)
+                                 errors() &= 16u;
```
Comment (tcclevenger) [inline containers/unit_tests/TestDynRankView_TeamScratch.hpp:51]: Don't you need `|=` here? I think you'd lose all the previous information in `error()` because `16u` has 0s in all those places.

Comment (crtrott) [reply containers/unit_tests/TestDynRankView_TeamScratch.hpp:51]: yeah

**Review by masterleinad — APPROVED** [review]
Looks OK to me.

**Comment by crtrott** [comment]
Retest this please.

File: containers/unit_tests/TestDynRankView_TeamScratch.hpp [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+#include <gtest/gtest.h>
+
+#include <Kokkos_DynRankView.hpp>
+
+namespace {
+
+void test_dyn_rank_view_team_scratch() {
+  using execution_space = TEST_EXECSPACE;
+  using memory_space    = execution_space::scratch_memory_space;
```
Comment (dalg24) [inline containers/unit_tests/TestDynRankView_TeamScratch.hpp:25]: not blocking but noting this is misleading

Comment (dalg24) [reply containers/unit_tests/TestDynRankView_TeamScratch.hpp:25]: (withdrawn)

**Review by dalg24 — APPROVED** [review]

**Comment by ndellingwood** [comment]
I patched these changes to the Trilinos pre-release PR for testing, https://github.com/trilinos/Trilinos/pull/13571

**Comment by ndellingwood** [comment]
This error is occurring in the Trilinos builds (AT2):
```
/home/Trilinos/build/packages/kokkos/containers/unit_tests/serial/TestSerial_DynRankView_TeamScratch.cpp:2:10: fatal error: TestDynRankView_TeamScratch.hpp: No such file or directory
 #include <TestDynRankView_TeamScratch.hpp>
```

I only cherry-picked this PR to the Trilinos tester PR, did I miss an auxiliary commit?

**Comment by tcclevenger** [comment]
@ndellingwood The cherry-picked commit 1b51cb835debf2cc207456dfd1df8fa06810659c in https://github.com/trilinos/Trilinos/pull/13571 for some reason did not add the unit test .hpp file, but added the entry in unit_tests/CMakeLists.txt, which is causing the error.

**Comment by tcclevenger** [comment]
@dalg24 Should I add this to the changelog under fixes?
